### PR TITLE
Add ItchySats App

### PR DESCRIPTION
**Pull request by klochowicz**
# App Submission

### App name
ItchySats

### Version
v0.3.0

### One line description of the app
Peer-2-peer derivatives on Bitcoin

### Summary of the app
ItchySats enables peer-2-peer CFD trading on Bitcoin using DLCs (discreet log contracts). No account needed, no trusted third-party - just you and your keys.

This is beta software. We tested it on test- and mainnet, but there are no guarantees that it will always behave as expected.
Please be mindful with how much money you trust the application with.
CFDs trading is inherently risky, be sure to read up on it before using this application.

That said: This is pretty awesome, go nuts!

1. Fund the ItchySats wallet
2. Open a position
3. Watch the price go up
4. Profit

Limitations of the mainnet beta:

1. You can only open long positions at the moment
2. Minimum position quantity is $100, maximum $1000
3. CFDs period ends after 7 days - perpetual positions are in the making :)
4. The leverage is fixed at 2

We are woking hard on perpetual positions and allowing sell positions.
Update to be expected soon!

### Developer name
ItchySats

### Developer website
https://itchysats.network

### Source code link
https://github.com/itchysats/itchysats

### Support link
https://github.com/itchysats/itchysats/issues

### Requires
- [ ] Bitcoin Core
- [x] Electrum server
- [ ] LND

### 256x256 SVG icon

[Itchilogo.svg.zip](https://github.com/getumbrel/umbrel/files/7665831/Itchilogo.svg.zip)
(as zip as svgur.com is over quota and does not work now...)

### Gallery images
(3 screenshots with captions)

"Start it. Trade! It’s that simple"

"Bitcoin CFDs without a middle-man"

"No accounts, just you and your keys!"

### I have tested my app on:
- [ ] [Umbrel dev environment](https://github.com/getumbrel/umbrel-dev)
- [x] [Umbrel OS on a Raspberry Pi 4](https://github.com/getumbrel/umbrel-os)
- [ ] [Custom Umbrel install on Linux](https://github.com/getumbrel/umbrel#-installation)

### CODE REVIEW

**Comment by itchymax** [comment]
It's a draft because we still have to release a new version :) Please have a look already and feel free to test it :)

File: apps/itchysats/docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    image: ghcr.io/itchysats/itchysats/taker:0.3.0@sha256:89adbae24d5c36c2516357b1267fc8f0c8fed59f99dc3c07b75617685ef525c5
+    restart: on-failure
+    stop_grace_period: 1m
+    ports:
+      - ${APP_ITCHYSATS_PORT}:8000
+    volumes:
+      - ${APP_DATA_DIR}/data:/data
+    command:
+      - --maker=mainnet.itchysats.network:9999
+      - --maker-id=7e35e34801e766a6a29ecb9e22810ea4e3476c2b37bf75882edf94a68b1d9607
```
Comment (aphex3k) [inline apps/itchysats/docker-compose.yml:14]: What is this id used for?

Comment (itchymax) [reply apps/itchysats/docker-compose.yml:14]: It's the maker's public key which is needed to have an encrypted connection with the maker behind `mainnet.itchysats.network:9999`.

**Comment by klochowicz** [comment]
I've force-pushed after rebasing against the recent umbrel `master` to avoid merge conflicts (no other changes).

File: scripts/configure [hunk 2]
```diff
@@ -218,6 +218,8 @@ APP_SYNCTHING_PORT="8384"
 APP_SYNCTHING_SYNC_PORT="22000"
 APP_UPTIME_KUMA_PORT="8385"
 APP_UPTIME_KUMA_IP="10.21.21.62"
+APP_ITCHYSATS_IP="10.21.21.63"
```
Comment (lukechilds) [inline scripts/configure:221]: There is an IP collision with `$APP_RIDE_THE_LIGHTNING_BOLTZ_IP` in master here.

```suggestion
APP_ITCHYSATS_IP="10.21.21.64"
```

**Review by lukechilds — CHANGES_REQUESTED** [review]
This looks great guys, some minor tweaks required.

Btw, I'm unable to test this on my mainnet node, when I try to install I get:

```
umbrel@umbrel-mainnet:~/umbrel $ ./scripts/app install itchysats
Setting up data dir for app itchysats...
sending incremental file list

sent 126 bytes  received 13 bytes  278.00 bytes/sec
total size is 654  speedup is 4.71
Pulling images for app itchysats...
Pulling web ... done
Starting app itchysats...
Creating itchysats_web_1 ... error

ERROR: for itchysats_web_1  Cannot start service web: OCI runtime create failed: invalid mount {Destination:data Type:bind Source:/var/lib/docker/volumes/3373facda23457bc501bfff215d91789cc1ee4f20c043d4b929bed56fc691305/_data Options:[rbind]}: mount destination data not absolute: unknown

ERROR: for web  Cannot start service web: OCI runtime create failed: invalid mount {Destination:data Type:bind Source:/var/lib/docker/volumes/3373facda23457bc501bfff215d91789cc1ee4f20c043d4b929bed56fc691305/_data Options:[rbind]}: mount destination data not absolute: unknown
ERROR: Encountered errors while bringing up the project.
```

I was however able to succesfully install on my testnet node, so seems like I'm hitting some strange Docker bug and not an issue with your app specifically. Just checking you were able to install ok?

Also are you able to allow edits from maintainers so I can push to this PR branch?

**Comment by lukechilds** [comment]
Also where's the best place to DM you guys for some more general questions on ItchySats?

**Comment by lukechilds** [comment]
Also what license is the ItchySats codebase licensed under? Currently I don't see a license file on your repo.

**Comment by lukechilds** [comment]
Seems like the volume issue is that this line https://github.com/itchysats/itchysats/blob/61c7bcbb322ab4adc43c199210a423287862a8d1/Dockerfile#L13 should be `VOLUME /data` (absolute instead of relative path).

Fixed here: https://github.com/itchysats/itchysats/pull/971

**Comment by lukechilds** [comment]
Hitting this trying to open a position:

Screenshot 2021-12-24 at 18 01 39

Is that expected currently if the app is still in testing and you don't have a lot of funds on the maker node or is something wrong?

Also seeing lots of stuff like this spamming the logs:

```
web_1  | 2021-12-24 11:09:36  WARN daemon::tokio_ext: Task failed: Failed to GET https://h00.ooo/x/BitMEX/BXBT/2021-12-26T12:00:00.price?n=20: error sending request for url (https://h00.ooo/x/BitMEX/BXBT/2021-12-26T12:00:00.price?n=20): error trying to connect: dns error: failed to lookup a
ddress information: Temporary failure in name resolution: error trying to connect: dns error: failed to lookup address information: Temporary failure in name resolution: dns error: failed to lookup address information: Temporary failure in name resolution: failed to lookup address informati
on: Temporary failure in name resolution
web_1  | 2021-12-24 11:09:36  WARN daemon::tokio_ext: Task failed: Failed to GET https://h00.ooo/x/BitMEX/BXBT/2021-12-31T03:00:00.price?n=20: error sending request for url (https://h00.ooo/x/BitMEX/BXBT/2021-12-31T03:00:00.price?n=20): error trying to connect: dns error: failed to lookup a
ddress information: Temporary failure in name resolution: error trying to connect: dns error: failed to lookup address information: Temporary failure in name resolution: dns error: failed to lookup address information: Temporary failure in name resolution: failed to lookup address informati
on: Temporary failure in name resolution
web_1  | 2021-12-24 11:09:36  WARN daemon::tokio_ext: Task failed: Failed to GET https://h00.ooo/x/BitMEX/BXBT/2021-12-26T15:00:00.price?n=20: error sending request for url (https://h00.ooo/x/BitMEX/BXBT/2021-12-26T15:00:00.price?n=20): error trying to connect: dns error: failed to lookup a
ddress information: Temporary failure in name resolution: error trying to connect: dns error: failed to lookup address information: Temporary failure in name resolution: dns error: failed to lookup address information: Temporary failure in name resolution: failed to lookup address informati
on: Temporary failure in name resolution
web_1  | 2021-12-24 11:09:36 DEBUG daemon::oracle: Fetching announcement for /x/BitMEX/BXBT/2021-12-30T11:00:00.price?n=20
web_1  | 2021-12-24 11:09:36 DEBUG daemon::oracle: Fetching announcement for /x/BitMEX/BXBT/2021-12-29T12:00:00.price?n=20
web_1  | 2021-12-24 11:09:36 DEBUG daemon::oracle: Fetching announcement for /x/BitMEX/BXBT/2021-12-29T13:00:00.price?n=20
web_1  | 2021-12-24 11:09:36 DEBUG daemon::oracle: Fetching announcement for /x/BitMEX/BXBT/2021-12-29T14:00:00.price?n=20
```

File: apps/registry.json [hunk 3]
```diff
@@ -657,5 +657,29 @@
...
+        "version": "v0.3.0",
+        "tagline": "Peer-2-peer derivatives on Bitcoin",
+        "description": "ItchySats enables peer-2-peer CFD trading on Bitcoin using DLCs (discreet log contracts). No account needed, no trusted third-party - just you and your keys.\n\nThis is beta software. We tested it on test- and mainnet, but there are no guarantees that it will always behave as expected.\nPlease be mindful with how much money you trust the application with.\nCFDs trading is inherently risky, be sure to read up on it before using this application.\n\nThat said: This is pretty awesome, go nuts!\n\n1. Fund the ItchySats wallet\n2. Open a position\n3. Watch the price go up\n4. Profit\n\nLimitations of the mainnet beta:\n\n1. You can only open long positions at the moment\n2. Minimum position quantity is $100, maximum $1000\n3. CFDs period ends after 7 days - perpetual positions are in the making :)\n4. The leverage is fixed at 2\n\nWe are woking hard on perpetual positions and allowing sell positions.\nUpdate to be expected soon!",
+        "developer": "ItchySats",
+        "website": "https://itchysats.network",
+        "dependencies": [
+            "bitcoind",
```
[the remark this answers is not in the document]
Comment (itchymax) [reply apps/registry.json]: 👍

File: templates/torrc-apps-3-sample [hunk 4]
```diff
@@ -48,3 +48,7 @@ HiddenServicePort 80 <app-syncthing-ip>:<app-syncthing-port>
 # uptime-kuma Hidden Service
 HiddenServiceDir /var/lib/tor/app-uptime-kuma
 HiddenServicePort 80 <app-uptime-kuma-ip>:3001
+
+# itchysats Hidden Service
+HiddenServiceDir /var/lib/tor/app-itchysats
+HiddenServicePort 80 <app-itchysats-ip>:<app-itchysats-port>
```
[the remark this answers is not in the document]
Comment (itchymax) [reply templates/torrc-apps-3-sample]: 👍

File: scripts/configure (hunk 2 above)
Comment (itchymax) [reply scripts/configure]: 👍 [`218b8dd` (#1149)](https://github.com/getumbrel/umbrel/pull/1149/commits/218b8dd021a8bea5817b3ee0e110fad98023f858)

**Comment by itchymax** [comment]
Hi @lukechilds 

thanks for reviewing our PR. I've fixed your comments in separate commits. 

Regarding the authentication: that's a good point. We do have a ticket for this https://github.com/itchysats/itchysats/issues/609 . Do you want this in the first release or is ok if we postpone it for the next release? 

> Also where's the best place to DM you guys for some more general questions on ItchySats?

You can find us on matrix: https://matrix.to/#/#comit-hermes:matrix.org or reach us via telegram https://t.me/joinchat/ULycH50PLV1jOTI0 

> Is that expected currently if the app is still in testing and you don't have a lot of funds on the maker node or is something wrong?

This looks correct and maybe we should better displayed the "No Liquidity" error a warning and not as an error: We are testing with safe amounts at the moment: $100-$1000 and the taker will create a new offer after a few seconds. 

The error below (Announcement not found) was not expected :/ It looks like there was a DNS error on your machine? 

> Also what license is the ItchySats codebase licensed under? Currently I don't see a license file on your repo.

Let me get back to you on that. Should be an open source license. 

p.s. merry christmas :)

**Comment by bonomat** [comment]
Hi @lukechilds, thanks for giving ItchySats a try. 

We have 3 versions running. As far as I remember, we started with version v0.4.8 but then upgraded to v0.4.9. All three version were running without any issues.

Regarding the DNS errors: it looks weird but I think we have seen this before. 
Can you open a ticket on our [repository](https://github.com/itchysats/itchysats/) and attach some more logs please? Maybe there is something useful in there before the lines you shared above. 
Also, it would be good to know if you had a "clean" umbrel node or if you changed something in the dns settings. 

Regarding authentication: I would opt for a nginx proxy using `APP_PASSWORD` for now until we get https://github.com/itchysats/itchysats/issues/609 implemented. This should be faster and we can then ship the authentication with a new release.

> Also are you able to allow edits from maintainers so I can push to this PR branch?

@klochowicz : please enable edits for maintainers

**Comment by lukechilds** [comment]
>p.s. merry christmas :)

Thanks! You too, hope you guys had a great Christmas!

>Regarding the authentication: that's a good point. We do have a ticket for this itchysats/itchysats#609 . Do you want this in the first release or is ok if we postpone it for the next release?
>-- @itchymax

To be honest it's up to you guys but I'd strongly advise adding some sort of basic auth to the initial release. Or atleast showing a big warning that this is an initial beta release with no auth if you decide not to.

This sounds like a great compromise though:

>Regarding authentication: I would opt for a nginx proxy using APP_PASSWORD for now until we get itchysats/itchysats#609 implemented. This should be faster and we can then ship the authentication with a new release.
>-- @bonomat

Re DNS

>The error below (Announcement not found) was not expected :/ It looks like there was a DNS error on your machine?
>-- @itchymax

>Regarding the DNS errors: it looks weird but I think we have seen this before.
>Can you open a ticket on our repository and attach some more logs please? Maybe there is something useful in there >before the lines you shared above.
>Also, it would be good to know if you had a "clean" umbrel node or if you changed something in the dns settings.
>-- @bonomat

I haven't changed any DNS settings. I just remembered some users had some issues with the mempool app that were resolved by changing Docker DNS settings. Sounds like this could be related. Let me try the [proposed fix](https://github.com/getumbrel/umbrel/issues/1076#issuecomment-991682514) and see if that resolves the ItchySats issue.

**Comment by lukechilds** [comment]
Also re this:

>We have 3 versions running. As far as I remember, we started with version v0.4.8 but then upgraded to v0.4.9. All three version were running without any issues.

Assuming that version is maker specific? Cos this PR currently includes `ghcr.io/itchysats/itchysats/taker:0.3.0`.

**Comment by bonomat** [comment]
> Also re this:
> 
> > We have 3 versions running. As far as I remember, we started with version v0.4.8 but then upgraded to v0.4.9. All three version were running without any issues.
> 
> Assuming that version is maker specific? Cos this PR currently includes `ghcr.io/itchysats/itchysats/taker:0.3.0`.

sorry, I meant Umbrel version.

**Comment by lukechilds** [comment]
>sorry, I meant Umbrel version.

Ah got it.

Btw just seen you have `ghcr.io/itchysats/itchysats/taker:0.3.2` published, we should update this PR to that, right?

**Comment by lukechilds** [comment]
Ok so this https://github.com/getumbrel/umbrel/issues/1076#issuecomment-991682514 fixed the DNS issue for me, so seems unrelated to ItchySats and some sort of internal Umbrel DNS issue. Looking into it.

Screenshot 2021-12-27 at 18 28 59

Opened a position! Exciting stuff 🚀

**Comment by bonomat** [comment]
Re 
> Btw just seen you have ghcr.io/itchysats/itchysats/taker:0.3.2 published, we should update this PR to that, right?

yes indeed: @itchymax / @klochowicz please

> Ok so this [#1076 (comment)](https://github.com/getumbrel/umbrel/issues/1076#issuecomment-991682514) fixed the DNS issue for me, so seems unrelated to ItchySats and some sort of internal Umbrel DNS issue. Looking into it.
> 
> Screenshot 2021-12-27 at 18 28 59
> 
> Opened a position! Exciting stuff 🚀

Great that it worked and let's go Bitcoin ! 🚀

**Comment by lukechilds** [comment]
Also guys, looks like you are generating the seed randomly on first start. This is fine, but we also expose `$APP_SEED` which is a hex hex string deterministically derived from the users Umbrel seed. If you wanted to you could make use of this. It would mean that if a user's Umbrel broke and they setup a new one and imported their previous Umbrel seed, the ItchySats app would then derive the same wallet as the previous install and reover the previous funds.

**Comment by itchymax** [comment]
> Also guys, looks like you are generating the seed randomly on first start. This is fine, but we also expose `$APP_SEED` which is a hex hex string deterministically derived from the users Umbrel seed. If you wanted to you could make use of this. It would mean that if a user's Umbrel broke and they setup a new one and imported their previous Umbrel seed, the ItchySats app would then derive the same wallet as the previous install and reover the previous funds.

That's cool. I've created a ticket for this: https://github.com/itchysats/itchysats/issues/978

**Comment by lukechilds** [comment]
@bonomat just FYI the DNS issue seems like a DNS issue on my network external to ItchySats and Umbrel.

Viewing https://outcome.observer/h00.ooo/x/BitMEX/BXBT in my browser on the same network gives this error:
Screenshot 2021-12-28 at 17 49 22

Changing to a different network or using Tor Browser allows me to view the price feed correctly. You said you had reports of this issue before, maybe there is some sort of problem with the DNS configuration of the price feed server causing it to not work with some ISPs? Any ideas?

**Comment by scratchscratchscratchy** [comment]
@lukechilds with upgrading to version `0.3.3` in https://github.com/getumbrel/umbrel/pull/1149/commits/64d6a05a1b64ff0b938f43c36f2d2c805a93f1e1 we restrict the API by basic auth (as discussed in comments above). Password is set to `$APP_PASSWORD`, username defaulted in registry to `itchysats`. 

I just tested upgrading my umbrel and it worked fine displaying the password/username in the appstore. 

I think with adding this we should be able to merge. Please let me know if there is anything else needed :)

**Comment by klochowicz** [comment]
@lukechilds @itchymax @bonomat Unfortunately I don't seem to be able to allow pushing to this branch as the PR was created from org repo instead of personal one. I can resubmit the PR from a personal github account to allow this, but I'm not sure if it's worth the effort?

**Comment by lukechilds** [comment]
@klochowicz ah yes, just seen this https://github.com/isaacs/github/issues/1681 that sucks.

Ok no worries, no need for a new PR, good to keep the history here, I can just submit a PR to your org repo if I need to do any changes.

**Comment by lukechilds** [comment]
>I think with adding this we should be able to merge. Please let me know if there is anything else needed :)

@scratchscratchscratchy awesome! Testing now!

**Review by lukechilds — APPROVED** [review]
Tested and working great. Amazing stuff guys, really excited to ship this!

**Pull request merged**